lib/vdc/DCRAM.cpp

```diff
@@ -0,0 +1,592 @@
+
+    _ncdfc->GetAtt(varname, attname, values);
+
+    return (true);
+}
+
+bool DCRAM::getAtt(string varname, string attname, vector<double> &values) const
+{
+    values.clear();
+
+    return (_getAttTemplate(varname, attname, values));
+}
+
+bool DCRAM::getAtt(string varname, string attname, vector<long> &values) const
+{
+    values.clear();
+
+    return (_getAttTemplate(varname, attname, values));
+}
+
+bool DCRAM::getAtt(string varname, string attname, string &values) const
+{
+    values.clear();
+
+    return (_getAttTemplate(varname, attname, values));
+}
+
+std::vector<string> DCRAM::getAttNames(string varname) const
+{
+    //    printf("%s(%s) = ", __func__, varname.c_str());
+
+    if (varname.empty()) {
+        return _ncdfc->GetAttNames("");
+    } else {
+        DC::BaseVar var;
+        bool        status = getBaseVarInfo(varname, var);
+        if (!status) {
+            //            printf("{} (ERR)\n");
+            return (vector<string>());
+        }
+
+        vector<string> names;
+
+        const std::map<string, Attribute> &         atts = var.GetAttributes();
+        std::map<string, Attribute>::const_iterator itr;
+        for (itr = atts.begin(); itr != atts.end(); ++itr) { names.push_back(itr->first); }
+
+        //        printf("%s\n", ToStr(names).c_str());
+        return (names);
+    }
+}
+
+DC::XType DCRAM::getAttType(string varname, string attname) const
+{
+    //    printf("%s(%s, %s)\n", __func__, varname.c_str(), attname.c_str());
+    DC::BaseVar var;
+    bool        status = getBaseVarInfo(varname, var);
+    if (!status) return (DC::INVALID);
+
+    DC::Attribute att;
+    status = var.GetAttribute(attname, att);
+    if (!status) return (DC::INVALID);
+
+    return (att.GetXType());
+}
+
+int DCRAM::getDimLensAtLevel(string varname, int, std::vector<size_t> &dims_at_level, std::vector<size_t> &bs_at_level) const
+{
+    VAssert(0);
+    return -1;
+}
+
+int DCRAM::getDimLensAtLevel(string varname, int, std::vector<size_t> &dims_at_level, std::vector<size_t> &bs_at_level, long ts) const
+{
+    dims_at_level.clear();
+    bs_at_level.clear();
+
+    bool ok = GetVarDimLens(varname, true, dims_at_level, ts);
+    if (!ok) {
+        SetErrMsg("Undefined variable name : %s", varname.c_str());
+        return (-1);
+    }
+
+    // Never blocked
+    //
+    bs_at_level = vector<size_t>(dims_at_level.size(), 1);
+
+    return (0);
+}
+
+
+int DCRAM::openVariableRead(size_t ts, string varname)
+{
+    //    printf("DCRAM::%s(%li, %s) = ", __func__, ts, varname.c_str());
+
+    int ret = _fakeVarsFileCounter++;
+    _fdMap[ret] = varname;
+    return ret;
+}
+
+
+int DCRAM::closeVariable(int fd)
+{
+    //    printf("DCRAM::%s(%i)\n", __func__, fd);
+
+    auto fdIt = _fdMap.find(fd);
+    if (fdIt != _fdMap.end()) _fdMap.erase(fdIt);
+
+    if (fd >= _fakeVarsFileCounterStart) {
+        _fakeVarsFileCounter--;
+        return 0;
+    }
+
+    VAssert(0);
```

Comment:
VAssert(0) and return(-1)?

Reply:
This is a failure state but we still need a valid return. I changed it to a regular `assert`.